src/Featurizers/PCAFeaturizer.h

```diff
@@ -0,0 +1,267 @@
+// |
+// ----------------------------------------------------------------------
+template <typename MatrixT>
+PCATransformer<MatrixT>::PCATransformer(MatrixT eigenvalues, MatrixT eigenvectors) :
+    _eigenvalues(
+        std::move(
+            [&eigenvalues](void) -> MatrixT & {
+                if(eigenvalues.size() == 0)
+                    throw std::invalid_argument("eigenvalues");
+
+                return eigenvalues;
+            }()
+        )
+    ),
+    _eigenvectors(
+        std::move(
+            [&eigenvectors](void) -> MatrixT & {
+                if(eigenvectors.size() == 0)
+                    throw std::invalid_argument("eigenvectors");
+
+                return eigenvectors;
+            }()
+        )
+    ) {
+}
+
+template <typename MatrixT>
+PCATransformer<MatrixT>::PCATransformer(Archive &ar) :
+    PCATransformer(
+        [&ar](void) {
+            // Version
+            std::uint16_t                   majorVersion(Traits<std::uint16_t>::deserialize(ar));
+            std::uint16_t                   minorVersion(Traits<std::uint16_t>::deserialize(ar));
+
+            if(majorVersion != 1 || minorVersion != 0)
+                throw std::runtime_error("Unsupported archive version");
+
+            // Data
+            MatrixT                        eigenvalues(Traits<MatrixT>::deserialize(ar));
+            MatrixT                        eigenvectors(Traits<MatrixT>::deserialize(ar));
+
+            return PCATransformer<MatrixT>(std::move(eigenvalues), std::move(eigenvectors));
+        }()
+    ) {
+}
+
+template <typename MatrixT>
+bool PCATransformer<MatrixT>::operator==(PCATransformer const &other) const {
+    if ((this->_eigenvalues - other._eigenvalues).norm() > 0.000001f || (this->_eigenvectors - other._eigenvectors).norm() > 0.000001f)
+        return false;
+
+    return true;
+}
+
+template <typename MatrixT>
+void PCATransformer<MatrixT>::save(Archive &ar) const /*override*/ {
+    // Version
+    Traits<std::uint16_t>::serialize(ar, 1); // Major
+    Traits<std::uint16_t>::serialize(ar, 0); // Minor
+
+    // Data
+    Traits<decltype(_eigenvalues)>::serialize(ar, _eigenvalues);
+    Traits<decltype(_eigenvectors)>::serialize(ar, _eigenvectors);
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  PCAEstimator
+// |
+// ----------------------------------------------------------------------
+template <typename MatrixT, size_t MaxNumTrainingItemsV>
+PCAEstimator<MatrixT, MaxNumTrainingItemsV>::PCAEstimator(AnnotationMapsPtr pAllColumnAnnotations, size_t colIndex) :
+    BaseType("SVDEstimatorImpl", std::move(pAllColumnAnnotations)),
+    _colIndex(
+        std::move(
+            [this, &colIndex](void) -> size_t & {
+                if(colIndex >= this->get_column_annotations().size())
+                    throw std::invalid_argument("colIndex");
+
+                return colIndex;
+            }()
+        )
+    ) {
+        _hasCalledFit = false;
```

Comment:
>_hasCalledFit [](start = 8, length = 13)

Initialize with the other member vars rather than inside the body of the constructor.

Reply:
sure